Improve the documentation for the 0.0.4 release

steve-s opened the pull request.

fangerer commented on docs/overview.rst:
```diff
@@ -200,9 +203,9 @@ API:
-will get the benefits of HPy automatically.
+If you use Cython, you can't use HPy directly. There is a work in progress to
+add Cython backend which emits HPy code instead of Python/C code: once this is
+done, you will get the benefits of HPy automatically.
```
of course, it does: https://github.com/cython/cython/pull/4490

steve-s replied:
done

steve-s commented:
> Is there a link to see how it will look like after rendering, or the only way to see it is to build the doc locally?

I don't know of any other way, but I am not very familiar with the deployment process of our docs.. :-)

I think I addressed the comment, so this is now ready for another review round.

fangerer commented:
Afaik, we don't deploy the docs built during the CI run. But to be honest, I'm also not fully aware of the process how it is deployed. I think @hodgestar knows best.
Anyway, you should be able to build the docs locally (at least I managed to do so once). See: https://github.com/hpyproject/hpy/blob/master/.github/workflows/ci.yml#L208

hodgestar commented:
The docs are build by readthedocs (see the `.readthedocs.yml` file and https://readthedocs.org/projects/hpy/. If you give me your readthedocs username I can add you as a maintainer.

Adding a documentation build to our CI is probably a good idea -- it's a better way to pick up documentation build errors quickly.

mattip commented:
> Adding a documentation build to our CI is probably a good idea

There is [a documentation build](https://github.com/hpyproject/hpy/blob/f941e8f530d4fe67be8606b6e01c69c52ef56fc6/.github/workflows/ci.yml#L209). It does not currently store the artifacts. There are ways to do that, for instance pypy/pypy.org uses the netlify github integrasion.

hodgestar commented:
Thanks @mattip.

We can also use GitHub directly to store the built docs, by adding, e.g.:
```yaml
      - name: Upload built HTML files
        uses: actions/upload-artifact@v2
        with:
          name: hpy_html_docs
          path: docs/_build/html/*
          if-no-files-found: error
```
to the `build_docs` job.

mattip commented:
> We can also use GitHub directly to store the built docs

This will upload a zip file of the artifact. Not quite a one-click preview, but maybe worthwhile

steve-s commented:
@hodgestar I am "stevesindelar" on readthedocs.org

steve-s commented on docs/api.rst:
```diff
@@ -19,6 +19,28 @@ Similarly, if you need a new handle for an existing object, you can duplicate
 
+The HPy API strictly follows these rules:
+
+- ``HPy`` handles are never borrowed, i.e., ``HPy`` handle returned by a API
```
done

steve-s commented on docs/debug-mode.rst:
```diff
@@ -22,7 +22,11 @@ The debugging context can already check for:
+* Usage of memory returned from API functions such as `HPyUnicode_AsUTF8AndSize`
+  after the corresponding handle was closed.
+* Write access to memory returned from API functions such as `HPyUnicode_AsUTF8AndSize`
+  that should be read-only.
```
done (I took the liberty to rephrase it a little bit)

steve-s commented on docs/overview.rst:
```diff
@@ -238,15 +241,23 @@ Current status and roadmap
-already in place. As on April 2021, the following milestones have been reached:
+already in place. As on April 2022, the following milestones have been reached:
+
+  - few real-world Python packages have been ported to HPy API.
```
done

steve-s commented on docs/porting-guide.rst:
```diff
@@ -30,30 +37,51 @@ variable which is unknown to the GC but contains a pointer to a GC-managed
+**IMPORTANT**: the contract of ``tp_traverse`` is that it must visit all the
+``HPyFields`` contained within given struct, or more precisely "owned" by given
+Python object (in the sense of the "owner" argument to ``HPyField_Store``), and
+nothing more or less. Some Python implementations may choose to not call the
```
done

hodgestar commented:
@steve-s Added you.

steve-s commented on hpy/debug/src/autogen_debug_ctx_init.h:
```diff
@@ -157,8 +157,8 @@ void debug_ctx_Tracker_ForgetAll(HPyContext *dctx, HPyTracker ht);
 DHPy debug_ctx_Field_Load(HPyContext *dctx, DHPy source_object, HPyField source_field);
-HPyThreadState debug_ctx_LeavePythonExecution(HPyContext *dctx);
 void debug_ctx_ReenterPythonExecution(HPyContext *dctx, HPyThreadState state);
+HPyThreadState debug_ctx_LeavePythonExecution(HPyContext *dctx);
```
Ok, so it won't be that simple, because we have a CI job that checks that someone did not forget to `make autogen` before pushing, which is think is a good thing in general. I am not so sure about alphabetical or some other artificial order. The order as it is has some logic to it, maybe we want to preserve it even in generated header, which some people may still want to inspect manually..

antocuni replied:
But if I read the diff correctly, you didn't change the ordering inside `public_api.h`, did you?
So, it means that the order inside the autogen files is basically random, and anything seems better than random to me.
But anyway, I think that this should not block merging the PR

mattip commented:
It seems the owners on https://readthedocs.org/projects/hpy/ are @antocuni, @hodgestar, @steve-s, one of them would need to enable the readthedocs intgration.

antocuni commented:
I have enabled the relevant setting in readthedocs. If anyone wants to be added to the maintainers, just ping me with your username
